beginner_source/aws_distributed_training_tutorial.py

```diff
@@ -0,0 +1,675 @@
+#    we are using several multi-gpu nodes, NCCL is suggested.
+# 
+# -  **dist\_url** - URL to specify the initialization method of the
+#    process group. This may contain the IP address and port of the rank0
+#    process or be a non-existant file on a shared file system. Here,
+#    since we do not have a shared file system this will incorporate the
+#    **node0-privateIP** and the port on node0 to use.
+# 
+
+print("Collect Inputs...")
+
+# Batch Size for training and testing
+batch_size = 32
+
+# Number of worker threads for dataloading
+workers = 2
+
+# Number of epochs to train for
+num_epochs = 2
+
+# Starting Learning Rate
+starting_lr = 0.1
+
+# Number of distributed processes
+world_size = 4
+
+# Distributed backend type
+dist_backend = 'nccl'
+
+# Url used to setup distributed training
+dist_url = "tcp://172.31.22.234:23456"
+
+
+######################################################################
+# Initialize process group
+# ~~~~~~~~~~~~~~~~~~~~~~~~
+# 
+# One of the most important parts of distributed training in PyTorch is to
+# properly setup the process group, which is the **first** step in
+# initializing the ``torch.distributed`` package. To do this, we will use
+# the ``torch.distributed.init_process_group`` function which takes
+# several inputs. First, a *backend* input which specifies the backend to
+# use (i.e. NCCL, Gloo, MPI, etc.). An *init\_method* input which is
+# either a url containing the address and port of the rank0 machine or a
+# path to a non-existant file on the shared file system. Note, to use the
+# file init\_method, all machines must have access to the file, similarly
+# for the url method, all machines must be able to communicate on the
+# network so make sure to configure any firewalls and network settings to
+# accomodate. The *init\_process\_group* function also takes *rank* and
+# *world\_size* arguments which specify the rank of this process when run
+# and the number of processes in the collective, respectively.
+# 
+# Another important step, especially when each node has multiple gpus is
+# to set the *local\_rank* of this process. For example, if you have two
+# nodes, each with 8 GPUs and you wish to train with all of them then
+# :math:`world\_size=16` and each node will have a process with local rank
+# 0-7. This local\_rank is used to set the device (i.e. which GPU to use)
+# for the process and later used to set the device when creating a
+# distributed data parallel model. It is also recommended to use NCCL
+# backend in this hypothetical environment as NCCL is preferred for
+# multi-gpu nodes.
+# 
+
+print("Initialize Process Group...")
+# Initialize Process Group
+# v1 - init with url
+dist.init_process_group(backend=dist_backend, init_method=dist_url, rank=int(sys.argv[1]), world_size=world_size)
+# v2 - init with file
+#dist.init_process_group(backend="nccl", init_method="file:///home/ubuntu/pt-distributed-tutorial/trainfile", rank=int(sys.argv[1]), world_size=world_size)
+
```

Comment:
v3 - init with env://, and please provide what the environmental variables should look like in the comment too.